test-apps/display-test-app/android/imodeljs-test-app/app/src/main/java/com/bentley/imodeljs_test_app/DtaServiceAuthorizationClient.kt

```diff
@@ -0,0 +1,114 @@
+import com.bentley.itwin.AuthorizationClient
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.json.JSONObject
+import java.io.BufferedReader
+import java.io.DataOutputStream
+import java.net.URL
+import java.util.*
+import javax.net.ssl.HttpsURLConnection
+
+class DtaServiceAuthorizationClient(env: JSONObject): AuthorizationClient() {
+    data class AuthSettings(val clientId: String, val clientSecret: String, val scope: String, val authority: String)
+    private val authSettings = parseEnv(env)
+    private var accessToken: String? = null
+    private var expirationDate: String? = null
+
+    companion object {
+        fun create(env: JSONObject): DtaServiceAuthorizationClient? {
+            return try {
+                DtaServiceAuthorizationClient(env)
+            } catch (ex: Exception) {
+                null
+            }
+        }
+    }
+    private fun parseEnv(env: JSONObject): AuthSettings {
+        val clientId = env.optStringNotEmpty("IMJS_OIDC_CLIENT_ID")
+        val clientSecret = env.optStringNotEmpty("IMJS_OIDC_CLIENT_SECRET")
+        val scope = env.optStringNotEmpty("IMJS_OIDC_SCOPE")
+        val authority = env.optStringNotEmpty("IMJS_SERVICE_AUTHORITY") ?: "https://ims.bentley.com"
+        if (clientId == null || clientSecret == null || scope == null) {
+            throw Exception("env does not contain necessary config data.")
+        }
+        val scopes = scope.split(" ")
+        if (scopes.contains("openid") || scopes.contains("email") || scopes.contains("profile") || scopes.contains("organization")) {
+            throw Exception("DtaServiceAuthSettings: Scopes for a service cannot include 'openid email profile organization'")
+        }
+        return AuthSettings(clientId, clientSecret, scope, authority)
+    }
+
+    private suspend fun generateAccessToken() {
+        return withContext(Dispatchers.IO) {
+            val authURL = URL("${authSettings.authority}/connect/token")
+            with (authURL.openConnection() as HttpsURLConnection) {
+                useCaches = false
+                doOutput = true
+                requestMethod = "POST"
+                val encodedScope = Uri.encode(authSettings.scope)
+                val body = "grant_type=client_credentials&scope=${encodedScope}"
+                val bodyData = body.toByteArray()
+                setRequestProperty("Accept", "application/json")
+                setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
+                setRequestProperty("charset", "utf-8")
+                setRequestProperty("Content-Length", "${bodyData.size}")
+                val encodedClientId = Uri.encode(authSettings.clientId)
+                val encodedClientSecret = Uri.encode(authSettings.clientSecret)
+                val encoded = "${encodedClientId}:${encodedClientSecret}"
```

Comment:
```suggestion
                val encoded = "${Uri.encode(authSettings.clientId)}:${Uri.encode(authSettings.clientSecret)}"
```

Reply:
I prefer what I have, as it is easier for me to read.